samples/AvatarWithRoundedCorner/Program.cs

```diff
@@ -8,62 +8,74 @@ namespace AvatarWithRoundedCorner
     using SixLabors.Primitives;
     using SixLabors.Shapes;
 
-    class Program
+    static class Program
     {
         static void Main(string[] args)
         {
             System.IO.Directory.CreateDirectory("output");
+            using (var img = Image.Load("fb.jpg"))
+            {
+                // as generate returns a new IImage make sure we dispose of it
+                using (Image<Rgba32> dest = img.Clone(x => x.ConvertToAvatar(new Size(200, 200), 20)))
+                {
+                    dest.Save("output/fb.png");
+                }
 
-            GenerateAvatar("fb.jpg", "output/fb.png", new Size(200, 200), 20);
-            GenerateAvatar("fb.jpg", "output/fb-round.png", new Size(200, 200), 100);
-            GenerateAvatar("fb.jpg", "output/fb-rounder.png", new Size(200, 200), 150);
+                using (Image<Rgba32> destRound = img.Clone(x => x.ConvertToAvatar(new Size(200, 200), 100)))
+                {
+                    destRound.Save("output/fb-round.png");
+                }
+
+                using (Image<Rgba32> destRound = img.Clone(x => x.ConvertToAvatar(new Size(200, 200), 150)))
+                {
+                    destRound.Save("output/fb-rounder.png");
+                }
+                
+                // the original `img` object has not been altered at all.
+            }
         }
 
-        private static void GenerateAvatar(string source, string destination, Size size, float cornerRadius)
+        // lets create our custom image mutating pipeline
+        private static IImageProcessingContext<Rgba32> ConvertToAvatar(this IImageProcessingContext<Rgba32> operations, Size size, float cornerRadius)
         {
-            using (var image = Image.Load(source))
+            return operations.Resize(new ImageSharp.Processing.ResizeOptions
             {
-                image.Resize(new ImageSharp.Processing.ResizeOptions
-                {
-                    Size = size,
-                    Mode = ImageSharp.Processing.ResizeMode.Crop
-                });
-
-                ApplyRoundedCourners(image, cornerRadius);
-                image.Save(destination);
-            }
+                Size = size,
+                Mode = ImageSharp.Processing.ResizeMode.Crop
+            }).Apply(i => ApplyRoundedCorners(i, cornerRadius));
```

Comment:
`CurrentImage` is exactly what we shouldn't have, adding that would prevent us easily being able to do optimizations around rolling up similar processors etc into single set of pixel operations or even things like detecting that some operations will have no effect on the final image thus can be skipped (eg after a user applies a lot of operations then calls `Fill()` with a solid color thus we could skip all the previous operations and just call fill as it'll fully set every pixel anyway.)

Or for rolling up we could have multiple calls to draw text we would be able to roll all the calls into a single operation that only renders the glyphs as a single pass instead of multiple operations.